Add restrict_accessor example

mhoemmen opened the pull request:
This example shows how to write a custom mdspan accessor that uses the `restrict` (or `__restrict` or `__restrict__`) attribute to optimize unaliased array access.  It comes from my comments on Issue https://github.com/kokkos/mdspan/issues/169 .

mhoemmen committed:
Add restrict_accessor example

This answers a common question about custom accessors.
It comes from my comments on Issue
https://github.com/kokkos/mdspan/issues/169 .

crtrott commented on examples/restrict_accessor/restrict_accessor.cpp:
```diff
@@ -0,0 +1,298 @@
+    return p[i];
+  }
+  constexpr typename offset_policy::data_handle_type
+  offset(data_handle_type p, size_t i) const noexcept {
```
Actually an interesting question: should we throw/abort in here? Because getting an alias to a "I will not be aliased during my lifetime" thing seems iffy?

mhoemmen replied:
@crtrott An excellent question!  My view about offset is that it should behave as the language behaves.  It's legal in C (and extensions of C with the `restrict` keyword or attribute) to assign from a `restrict` pointer to a non-`restrict` pointer.  It's meaningful and necessary, because of code that accesses the elements of a `restrict` pointer-to-array using a pointer.  For example:
```c++
void assign_to_elements_of_x(int* restrict x, int n) {
  int* beg = x;
  int* end = x+n;
  while(beg != end) {
    *beg++ = 42;
  }
}
```

mhoemmen replied:
Another example:
```c++
void compute_x(int* restrict x, int n) {
  span<int> x_whole{x, n};
  base_case(x_whole);
  span<int> x_left{x, n/2};
  span<int> x_right{x + n/2, n/2};
  compute_left(x_left);
  compute_right(x_right);
}
```

crtrott reviewed (commented):
Generally like this.

mhoemmen committed:
Fix tiled layout example

Ensure that the tiled layout mapping
meets the latest layout mapping requirements.

dalg24 commented on examples/restrict_accessor/restrict_accessor.cpp:
```diff
@@ -0,0 +1,298 @@
+}
+
+template<class ElementType>
+void warmup(const index_type n, const ElementType* x, const ElementType* y, ElementType* z)
+{
+  constexpr std::size_t num_warmup_trials = 10;
+  for(std::size_t trial = 0; trial < num_warmup_trials; ++trial) {
+    for (index_type i = 0; i < n; ++i) {
+      z[i] = x[i] + y[i];
+    }
+  }
```
Any good reason not to just the benchmarks directly to do the warmup?

mhoemmen replied:
@dalg24 lol no ; - )

mhoemmen replied:
@dalg24 fixed; thanks!

mhoemmen committed:
restrict_accessor: remove code duplication

warmup function had duplicated code; this commit fixes that.
